### PR TITLE
Dockerfile: update ALPINE_VERSION to 3.18

### PR DESCRIPTION
**Pull request by thaJeztah**
- relates to https://github.com/docker/cli/pull/4350
- [x] depends on https://github.com/docker/cli/pull/4576
- [x] depends on https://github.com/docker/cli/pull/4577

This also moves `musl-dev` to the alpine-base stage, due to changes in
Alpine 3.18 causing gotestsum build to fail because stdlib.h was missing;

    #17 5.065 # runtime/cgo
    #17 5.065 In file included from _cgo_export.c:3:
    #17 5.065 /usr/include/fortify/stdlib.h:23:15: fatal error: stdlib.h: No such file or directory
    #17 5.065    23 | #include_next <stdlib.h>
    #17 5.065       |               ^~~~~~~~~~

alpine 3.17:

    / # find / | grep stdlib.h
    /usr/include/c++/12.2.1/tr1/stdlib.h
    /usr/include/c++/12.2.1/stdlib.h

alpine 3.18

    / # find / | grep stdlib.h
    /usr/lib/llvm16/lib/clang/16/include/__clang_hip_stdlib.h
    /usr/include/fortify/stdlib.h
    /usr/include/c++/12.2.1/tr1/stdlib.h
    /usr/include/c++/12.2.1/stdlib.h

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Ah, found the culprit; it's not Go, but the alpine upgrade;

```

**Comment by thaJeztah** [comment]
Looks like alpine 3.17 still works, and latest golang also has a variant for 3.17, so let's pick that for now

**Comment by thaJeztah** [comment]
It looks like there's a couple of options;

### remove `clang` for building `gotestsum`

it looks like `gotestsum` doesn't need it, and the error only happens if `clang` is installed (we should try to understand _why_ that changed in alpine 3.18 / newer clang?). I _think_ we can move `clang` from `build-base-alpine` to `build-alpine`; https://github.com/docker/cli/blob/7d723e2ba7b8607cf9ca5b3ea541f1ce60c700ec/Dockerfile#L13-L21

### add `musl-dev` to `build-base-alpine`

Adding `musl-dev` brings the header


### Reproduce

Here's a minimal dockerfile and build-args to test the variants;

```dockerfile
# syntax=docker/dockerfile:1

ARG ALPINE_VERSION=3.17
ARG BASE=default

FROM golang:1.20-alpine${ALPINE_VERSION} AS base-no-clang

FROM base-no-clang AS base-default
RUN apk add --no-cache clang

FROM base-default AS base-clang-musl-dev
RUN apk add --no-cache musl-dev

FROM base-${BASE}
ARG GOTESTSUM_VERSION=v1.8.2
RUN --mount=type=cache,target=/go/pkg/mod \
    go install "gotest.tools/gotestsum@${GOTESTSUM_VERSION}" && gotestsum --version
```


| x                  | ALPINE_VERSION | clang?             | musl-dev?          | cmd                                                                              |
|--------------------|----------------|--------------------|--------------------|----------------------------------------------------------------------------------|
| :white_check_mark: | 3.17           | -                  | -                  | `docker build --build-arg BASE=no-clang .`                                       |
| :white_check_mark: | 3.17           | :white_check_mark: | -                  | `docker build .`                                                                 |
| :white_check_mark: | 3.17           | :white_check_mark: | :white_check_mark: | `docker build --build-arg BASE=clang-musl-dev .`                                 |
| :white_check_mark: | 3.18           | -                  | -                  | `docker build --build-arg ALPINE_VERSION=3.18 --build-arg BASE=no-clang .`       |
| :x:                | 3.18           | :white_check_mark: | -                  | `docker build --build-arg ALPINE_VERSION=3.18  .`                                |
| :white_check_mark: | 3.18           | :white_check_mark: | :white_check_mark: | `docker build --build-arg ALPINE_VERSION=3.18 --build-arg BASE=clang-musl-dev .` |

**Comment by thaJeztah** [comment]
From a discussion on Slack;

after installing clang

alpine 3.17:

```bash
/ # find / | grep stdlib.h
/usr/include/c++/12.2.1/tr1/stdlib.h
/usr/include/c++/12.2.1/stdlib.h
```


alpine 3.18

```bash
/ # find / | grep stdlib.h
/usr/lib/llvm16/lib/clang/16/include/__clang_hip_stdlib.h
/usr/include/fortify/stdlib.h
/usr/include/c++/12.2.1/tr1/stdlib.h
/usr/include/c++/12.2.1/stdlib.h
```

**Comment by thaJeztah** [comment]
LOL, I was wondering why the e2e tests were failing, and was looking through logs, but couldn't find an actual failure there;

```bash
make -f docker.Makefile test-e2e-non-experimental

#6 [2/2] COPY ./notary/ /fixtures/
#6 CACHED

#7 exporting to image
#7 exporting layers done
#7 writing image sha256:caf0815eaa7da5cd65244aa32318f904951ed74e2ee2c3c5bad2af2ed299d85e done
#7 naming to docker.io/library/cliendtoendsuite-notary-server done
#7 DONE 0.0s
 Container cliendtoendsuite-notary-server-1  Running
 Container cliendtoendsuite-evil-notary-server-1  Running
 Container cliendtoendsuite-registry-1  Running
 Container cliendtoendsuite-engine-1  Running
Error: No such object: cliendtoendsuite_engine_1
make: *** [docker.Makefile:142: test-e2e-non-experimental] Error 1
```

And it was staring me in the face:

```
 Container cliendtoendsuite-engine-1  Running
Error: No such object: cliendtoendsuite_engine_1
```

Looks like something in the e2e tests is expecting compose to be creating containers using underscores, and we're installing docker-compose from alpine's package repositories; and it looks like alpine 3.18 switched over to use compose v2:


On alpine 3.17:

```bash
make -f docker.Makefile build-e2e-image
docker run --rm docker-cli-e2e docker-compose --version
docker-compose version 1.29.2, build unknown
```

On alpine 3.18:

```bash
make -f docker.Makefile build-e2e-image
docker run --rm docker-cli-e2e docker-compose --version
Docker Compose version v2.17.3
```

File: Dockerfile [hunk 1]
```diff
@@ -14,13 +14,13 @@ FROM --platform=$BUILDPLATFORM tonistiigi/xx:${XX_VERSION} AS xx
...
 COPY --link --from=xx / /
-RUN apk add --no-cache bash clang lld llvm file git
+RUN apk add --no-cache bash clang lld llvm musl-dev file git
 WORKDIR /go/src/github.com/docker/cli
 
 FROM build-base-alpine AS build-alpine
 ARG TARGETPLATFORM
 # gcc is installed for libgcc only
-RUN xx-apk add --no-cache musl-dev gcc
+RUN xx-apk add --no-cache gcc
```
Comment (thaJeztah) [inline Dockerfile]: I worked around this by not using `cgo` for `gotestsum`, but wondering if we still need this, or something similar, but not sure if `musl-dev` must be installed through `xx` here, or if it's safe to install it using a plain `apk add`

/cc @crazy-max perhaps you have ideas   her

**Comment by thaJeztah** [comment]
Alright, this one is happy now as well; no longer had to change the `musl-dev` parts

/cc @crazy-max ptal 🤗

**Review by crazy-max — APPROVED** [review]

**Pull request merged**